eclipse-theia/theia
# 16502 plugin startup

#### What it does
This is a fix for https://github.com/eclipse-theia/theia/issues/8429. There are a couple of changes and refactorings going on here:

* In order to make sure that startup of plugin hosts happens in a well defined sequence, plugin hosts announce that they are starting by sending `willStartPluginHost`. Once they have finished, they send `didStartPluginHost` to the `HostedPluginClient` on the browser side.
* ✅ `MessageConnection` and related code always sends strings around. It is the responsibility of clients to `stringify` and `parse` objects
* ✅ `RPCMultiplexer` is now an implementation detail of `RPCProtocol`. 
* ✅ Sending messages between plugin hosts and clients now requires an explicit `pluginHostId`. So far, there was an implicit protocol based on `pluginId`, which just took the first plugin deployed to a plugin host to do routing. 
* ✅ Rename `HostedPluginServer/Client` in the plugin-dev package to `PluginDevServer/Client`.  Before, the wrong client interface was used for the plugin dev server, showing that the naming was confusing before.
* Turned `HostedPluginRunner` into just another `ServerPluginRunner`, removing the special case.

#### How to test
Install various plugins and check what happens to them 
* upon starting up for the first time
* kill the back end and start it again: plugins should come back correctly
* Refresh the browser window: plugins should be shut down and restarted.

Note: front-end plugins are currently borked and therefore could not be tested (https://github.com/eclipse-theia/theia/issues/6601)

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
